pinterest/teletraan
# Update agent record if rollback deploy id is transformed.

Problem:
--
Sometimes deploy gets stuck when rollbacked. This issue does not happen all the times however, I am able to repro this fairly consistently with following sequence: 1. Env. is currently on build X 2. Deploy build Y 3. Rollback to X while Y is in progress. 4. Observe some host which are not on Y yet, get stuck forever. 
This happens because the stuck hosts are already on build X. There is no impact on service uptime however, deploy get's stuck.

Fix: 
--
Teletraan agent service transforms agent's deploy id for a given environment to the rollback deploy id as host is already on to this deploy (ex: X above). Earlier we used to update agents record regardless of a change in record from ping request handling. Now, we update only if there is a change in agent record from ping request handling. Since, here agent service transforms deploy id with rollback deploy id, we need to update agent record in the db to reflect that.

Test: 
--
1. Verify deploy success
2. Verify rollback success while deploy is in progress
3. Verify rollback success once deploy is complete
